File: web/src/game/upgrade.ts
// src/game/upgrade.ts
// UpgradeSystem: skill catalog + skill tree branches. Mirrors desktop
// upgrade.py: every level grants a skill point, spent in the SKILL TREE
// (Combat / Survival / Utility branches) instead of a random 3-card pick.

import type { UpgradeCatalog } from "./data";
import type { IGame } from "./types";

/** Skill-tree branch definition: name, accent colour, skill ids. */
export const SKILL_BRANCHES: ReadonlyArray<{
  name: string;
  color: string;
  skills: readonly string[];
}> = [
  {
    name: "Combat",
    color: "#FFC850",
    skills: ["damage", "fire_rate", "reload", "crit_ch", "crit_dmg"],
  },
  {
    name: "Survival",
    color: "#6EDC82",
    skills: ["max_hp", "armor", "regen", "vampire"],
  },
  {
    name: "Utility",
    color: "#6EC8FF",
    skills: ["speed", "magnet", "pierce"],
  },
];

/**
 * Seconds the level-up card overlay ignores clicks after it opens. The player
 * is usually mid-firefight (mouse button held down) when a level lands, so the
 * first moments of the overlay must swallow those clicks instead of picking a
 * skill by accident. After it elapses the overlay waits indefinitely.
 */
export const LEVELUP_PICK_LOCK = 3;

/** Max levels for skills that have no explicit catalog limit. */
export const DEFAULT_SKILL_LIMIT = 5;

export class UpgradeSystem {
  catalog: UpgradeCatalog;

  constructor(catalog: UpgradeCatalog) {
    this.catalog = catalog;
  }

  limitFor(uid: string): number {
    return this.catalog.limits[uid] ?? DEFAULT_SKILL_LIMIT;
  }

  textFor(uid: string): string {
    return this.catalog.upgrades.find((u) => u.id === uid)?.text ?? uid;
  }

  descFor(uid: string): string {
    return this.catalog.upgrades.find((u) => u.id === uid)?.desc ?? "";
  }

  apply(uid: string, player: {
    upgradeLevels: Record<string, number>;
    maxHp: number;
    heal: (n: number) => void;
    addArmor: (n: number) => void;
    damageMult: number;
    speedMult: number;
    fireRateMult: number;
    reloadMult: number;
    critBonus: number;
    critMultBonus: number;
    regen: number;
    magnetMult: number;
    lifeSteal: number;
    pierceBonus: number;
  }, _game?: IGame): void {
    player.upgradeLevels[uid] = (player.upgradeLevels[uid] ?? 0) + 1;
    switch (uid) {
      case "max_hp":
        player.maxHp += 20;
        player.heal(20);
        break;
      case "damage":
        player.damageMult *= 1.1;
        break;
      case "speed":
        player.speedMult *= 1.08;
        break;
      case "fire_rate":
        player.fireRateMult *= 1.08;
        break;
      case "reload":
        player.reloadMult *= 0.9;
        break;
      case "armor":
        player.addArmor(10);
        break;
      case "crit_ch":
        player.critBonus += 0.05;
        break;
      case "crit_dmg":
        player.critMultBonus += 0.25;
        break;
      case "regen":
        player.regen += 1;
        break;
      case "magnet":
        player.magnetMult *= 1.3;
        break;
      case "vampire":
        player.lifeSteal += 0.02;
        break;
      case "pierce":
        player.pierceBonus += 1;
        break;
    }
  }
}

/** Branch that owns a skill id (used to colour the level-up cards). */
export function branchForSkill(uid: string):
  | (typeof SKILL_BRANCHES)[number]
  | undefined {
  return SKILL_BRANCHES.find((b) => b.skills.includes(uid));
}

/**
 * Roll the level-up offer: ONE random skill per skill-tree branch, so the
 * player always sees three cards (Combat / Survival / Utility). Skills already
 * at their level cap are excluded; if a whole branch is capped out, the slot is
 * back-filled from the remaining skills of the other branches so the offer
 * still has three cards while anything is left to learn.
 */
export function rollLevelUpChoices(
  levels: Record<string, number>,
  limitFor: (uid: string) => number,
  rand: () => number = Math.random,
): string[] {
  const pick = (pool: readonly string[]): string =>
    pool[Math.floor(rand() * pool.length) % pool.length]!;

  const chosen: string[] = [];
  const spare: string[] = [];
  for (const branch of SKILL_BRANCHES) {
    const pool = branch.skills.filter((uid) => (levels[uid] ?? 0) < limitFor(uid));
    if (pool.length === 0) continue;
    const uid = pick(pool);
    chosen.push(uid);
    for (const other of pool) if (other !== uid) spare.push(other);
  }
  while (chosen.length < SKILL_BRANCHES.length && spare.length > 0) {
    const i = Math.floor(rand() * spare.length) % spare.length;
    chosen.push(spare.splice(i, 1)[0]!);
  }
  return chosen;
}

File: upgrade.py
"""Level-up upgrade choices (pick 1 of 3) + Skill Tree branches.

Each level-up grants a `skill_point` instead of choosing from a 3-card pool.
Skill points can be spent in the SKILL TREE (menu → SKILLS) on three
branches: Combat (damage / fire rate), Survival (HP / armor / regen),
and Utility (speed / crit / pickup magnet). The classic 3-card choice
is kept as a fallback when SKILL_TREE_ENABLED is False.
"""
from __future__ import annotations

import random

import settings as S
from utils import load_json

_UPGRADES_FILE = S.UPGRADES_FILE if hasattr(S, "UPGRADES_FILE") \
    else S.DATA_DIR + "/upgrades.json"

_DATA = load_json(_UPGRADES_FILE, {
    "upgrades": [
        {"id": "max_hp",    "text": "+20 MAX HP",        "desc": "Increase max health"},
        {"id": "damage",    "text": "+10% DAMAGE",       "desc": "All weapons hit harder"},
        {"id": "speed",     "text": "+8% MOVE SPEED",    "desc": "Run faster"},
        {"id": "fire_rate", "text": "+8% FIRE RATE",     "desc": "Shoot faster"},
        {"id": "reload",    "text": "+10% RELOAD SPEED", "desc": "Reload quicker"},
        {"id": "armor",     "text": "+10 ARMOR",         "desc": "Absorb damage first"},
        {"id": "crit_ch",   "text": "+5% CRIT CHANCE",   "desc": "More critical hits"},
        {"id": "crit_dmg",  "text": "+25% CRIT DAMAGE",  "desc": "Crits hurt more"},
        {"id": "regen",     "text": "+1 HP / sec",       "desc": "Slowly regenerate health"},
        {"id": "magnet",    "text": "+30% pickup range", "desc": "Loot vacuum reaches further"},
        {"id": "vampire",   "text": "+2% life steal",    "desc": "Heal on every kill"},
        {"id": "pierce",    "text": "+1 pierce bullet",  "desc": "Crossbow pierces more enemies"},
    ],
    "limits": {"max_hp": 10, "armor": 10},
})

SKILL_TREE_ENABLED = True
SKILL_BRANCHES = {
    "Combat": ["damage", "fire_rate", "reload", "crit_ch", "crit_dmg"],
    "Survival": ["max_hp", "armor", "regen", "vampire"],
    "Utility": ["speed", "magnet", "pierce"],
}


class UpgradeSystem:
    """Catalog of player upgrades; rolls 3 random choices per level-up."""

    CATALOG: list[dict] = _DATA.get("upgrades", [])
    LIMITS: dict[str, int] = _DATA.get("limits", {})

    def roll_choices(self, player) -> list[str]:
        pool: list[str] = []
        for u in self.CATALOG:
            limit = self.LIMITS.get(u["id"])
            if limit is not None and player.upgrade_levels.get(u["id"], 0) >= limit:
                continue
            pool.append(u["id"])
        if len(pool) < 3:
            pool = pool + [u["id"] for u in self.CATALOG][:3 - len(pool)]
        return random.sample(pool, min(3, len(pool)))

    def text_for(self, uid: str) -> str:
        for u in self.CATALOG:
            if u["id"] == uid:
                return u["text"]
        return uid

    def apply(self, uid: str, player, game=None) -> None:
        player.upgrade_levels[uid] = player.upgrade_levels.get(uid, 0) + 1
        if uid == "max_hp":
            player.max_hp += 20
            player.heal(20)
        elif uid == "damage":
            player.damage_mult *= 1.10
        elif uid == "speed":
            player.speed_mult *= 1.08
        elif uid == "fire_rate":
            player.fire_rate_mult *= 1.08
        elif uid == "reload":
            player.reload_mult *= 0.90
        elif uid == "armor":
            player.add_armor(10)
        elif uid == "crit_ch":
            player.crit_bonus += 0.05
        elif uid == "crit_dmg":
            player.crit_mult_bonus += 0.25
        elif uid == "regen":
            player.regen += 1.0
        elif uid == "magnet":
            player.magnet_mult *= 1.30
        elif uid == "vampire":
            player.life_steal += 0.02
        elif uid == "pierce":
            player.pierce_bonus += 1
